src/ray/gcs/gcs_server/gcs_actor_manager.cc

```diff
@@ -0,0 +1,299 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "gcs_actor_manager.h"
+#include <ray/common/ray_config.h>
+
+#include <utility>
+
+namespace ray {
+namespace gcs {
+
+ClientID GcsActor::GetNodeID() const {
+  const auto &raylet_id_binary = actor_table_data_.address().raylet_id();
+  if (raylet_id_binary.empty()) {
+    return ClientID::Nil();
+  }
+  return ClientID::FromBinary(raylet_id_binary);
+}
+
+void GcsActor::UpdateAddress(const rpc::Address &address) {
+  actor_table_data_.mutable_address()->CopyFrom(address);
+}
+
+const rpc::Address &GcsActor::GetAddress() const { return actor_table_data_.address(); }
+
+WorkerID GcsActor::GetWorkerID() const {
+  const auto &address = actor_table_data_.address();
+  if (address.worker_id().empty()) {
+    return WorkerID::Nil();
+  }
+  return WorkerID::FromBinary(address.worker_id());
+}
+
+void GcsActor::UpdateState(rpc::ActorTableData::ActorState state) {
+  actor_table_data_.set_state(state);
+}
+
+rpc::ActorTableData::ActorState GcsActor::GetState() const {
+  return actor_table_data_.state();
+}
+
+ActorID GcsActor::GetActorID() const {
+  return ActorID::FromBinary(actor_table_data_.actor_id());
+}
+
+TaskSpecification GcsActor::GetCreationTaskSpecification() const {
+  const auto &task_spec = actor_table_data_.task_spec();
+  return TaskSpecification(task_spec);
+}
+
+const rpc::ActorTableData &GcsActor::GetActorTableData() const {
+  return actor_table_data_;
+}
+
+rpc::ActorTableData *GcsActor::GetMutableActorTableData() { return &actor_table_data_; }
+
+/////////////////////////////////////////////////////////////////////////////////////////
+GcsActorManager::GcsActorManager(boost::asio::io_context &io_context,
+                                 gcs::ActorInfoAccessor &actor_info_accessor,
+                                 gcs::GcsNodeManager &gcs_node_manager,
+                                 LeaseClientFactoryFn lease_client_factory,
+                                 rpc::ClientFactoryFn client_factory)
+    : actor_info_accessor_(actor_info_accessor),
+      gcs_actor_scheduler_(new gcs::GcsActorScheduler(
+          io_context, actor_info_accessor, gcs_node_manager,
+          [this](std::shared_ptr<GcsActor> actor) {
+            // When there are no available nodes to schedule the actor the
+            // gcs_actor_scheduler will treat it as failed and invoke this handler. In
+            // this case, the actor should be appended to the `pending_actors_` and wait
+            // for the registration of new node.
+            pending_actors_.emplace(std::move(actor));
+          },
+          [this](std::shared_ptr<GcsActor> actor) {
+            auto worker_id = actor->GetWorkerID();
```

Comment:
> Probably worthwhile to move this into a method of GcsActorManager, since there's significant logic here.

Okey, I will add a protected method `OnActorCreateSuccess` to do that.